src/usart.rs

```diff
@@ -452,41 +462,53 @@ impl Peripheral for pac::USART2 {
 ///
 /// Can be passed to [`USART::enable`] to configure the baud rate for a USART
 /// peripheral.
-pub struct BaudRate<'frg> {
-    _uartfrg: &'frg UARTFRG,
+pub struct BaudRate<'clock, CS: ClockSource> {
+    _clock: &'clock CS,
 
     /// USART Baud Rate Generator divider value
     ///
     /// See user manual, section 13.6.9.
     brgval: u16,
 }
 
-impl<'frg> BaudRate<'frg> {
+impl<'clock, CS> BaudRate<'clock, CS> where CS: ClockSource {
     /// Create a `BaudRate` instance
     ///
     /// Creates a `BaudRate` instance from two components: A reference to the
-    /// [`UARTFRG`] and the BRGVAL.
+    /// clock source and the BRGVAL.
     ///
-    /// The [`UARTFRG`] controls U_PCLK, the clock that is shared by all USART
-    /// peripherals. Please configure it before attempting to create a
-    /// `BaudRate`. By keeping a reference to it, `BaudRate` ensures that U_PCLK
-    /// cannot be changes as long as the `BaudRate` instance exists.
+    /// The clock source is shared by all USART peripherals. Please configure it
+    /// before attempting to create a `BaudRate`. By keeping a reference to it,
+    /// `BaudRate` ensures that the clock source cannot be changed as long as
+    /// the `BaudRate` instance exists.
     ///
     /// BRGVAL is an additional divider value that divides the shared baud rate
     /// to allow individual USART peripherals to use different baud rates. A
     /// value of `0` means that U_PCLK is used directly, `1` means that U_PCLK
     /// is divided by 2 before using it, `2` means it's divided by 3, and so on.
     ///
     /// Please refer to the user manual, section 13.3.1, for further details.
-    pub fn new(uartfrg : &'frg UARTFRG, brgval : u16) -> Self {
+    pub fn new(clock: &'clock CS, brgval : u16) -> Self {
         Self {
-            _uartfrg: uartfrg,
-            brgval  : brgval,
+            _clock: clock,
+            brgval: brgval,
         }
     }
 }
 
 
+/// Implemented for USART clock sources
+pub trait ClockSource {
```

Comment:
The clock source stuff is the same for i2c & spi as well, so it probably makes sense to put it in syscon

Reply:
Good point. I didn't want to make to too generic at this point, as my goal was to get something working quickly.